File: platform_adapter/openlambda/ol.py
import gc

from platform_adapter.interface import PlatformAdapter
import os
import time
import subprocess
import requests
import re
import shutil
import pandas as pd
import signal
import glob
import psutil
import threading
import csv
import logging
from flask import Flask, request, jsonify

# PSS is the most accurate metric for memory usage,
# I didn't use the RSS, as it doesn't exclude shared memory.
# I also didn't use memory.usage_in_bytes in cgroup, as it include kernel memory allocation.
def get_pss(pid):
    process = psutil.Process(int(pid))
    return process.memory_full_info().pss / 1024


def get_rss(pid):
    process = psutil.Process(int(pid))
    return process.memory_info().rss / 1024


def get_total_mem(base, MODE="PSS"):
    total_pss = 0
    if MODE == "CG":
        memory_current_path = os.path.join(base, "memory.current")
        with open(memory_current_path, "r") as file:
            mem_in_bytes = int(file.readline())
        return mem_in_bytes / 1024

    for cg_folder in glob.glob(os.path.join(base, "cg-*")):
        procs_file = os.path.join(cg_folder, "cgroup.procs")

        if os.path.exists(procs_file):
            with open(procs_file, "r") as f:
                pids = f.readlines()
            for pid in pids:
                pid = pid.strip()
                if MODE == "PSS":
                    total_pss += get_pss(pid)
                elif MODE == "RSS":
                    total_pss += get_rss(pid)
    return total_pss


# clear cache is important, it affects the performance
def clear_cache():
    try:
        with open('/proc/sys/vm/drop_caches', 'w') as f:
            f.write('1')
            f.write('2')
            f.write('3')
        print("Cache cleared successfully.")
    except Exception as e:
        print(f"Error clearing cache: {e}")


def read_pid_cgroup_memory(pid):
    def read_file(file_path):
        try:
            with open(file_path, 'r') as file:
                return file.read().strip()
        except IOError as e:
            return f"Error reading file {file_path}: {e}"

    cgroup_path = f"/proc/{pid}/cgroup"
    cgroup_info = read_file(cgroup_path)
    cgroup_path_components = cgroup_info.split(':')[-1].strip()
    cgroup_memory_path = f"/sys/fs/cgroup{cgroup_path_components}/memory.current"
    memory_usage = int(read_file(cgroup_memory_path)) / 1024
    return memory_usage


def read_cgroup_memory():
    # read memory usage from default-ol-sandboxes sandbox
    cgroup_path = "/sys/fs/cgroup/default-ol-sandboxes"
    mem_in_bytes = 0
    for cg_folder in glob.glob(os.path.join(cgroup_path, "cg-*")):
        mem = os.path.join(cg_folder, "memory.current")
        if os.path.exists(mem):
            with open(mem, "r") as f:
                mem_in_bytes += int(f.readline())
    return mem_in_bytes / 1024


metrics_lock = threading.Lock()


class OL(PlatformAdapter):
    def __init__(self):
        self.load_config("platform_adapter/openlambda/config.json")
        self.ol_dir = self.config["ol_dir"]

        self.metrics = []
        self.current_dir = os.path.dirname(os.path.abspath(__file__))

        self.app = Flask(__name__)
        self.app.add_url_rule('/fork', view_func=self.recv_fork, methods=['POST'])
        self.app.add_url_rule('/start', view_func=self.recv_start, methods=['POST'])
        self.collectT = None
        self.fork_list = []
        self.start_list = []
    def recv_fork(self):
        json_data = request.get_json()
        self.fork_list.append(json_data)
        if json_data:
            return jsonify({}), 200
        else:
            return jsonify({}), 400
    def recv_start(self):
        json_data = request.get_json()
        self.start_list.append(json_data)
        if json_data:
            return jsonify({}), 200
        else:
            return jsonify({}), 400


    def start_worker(self, options={}):
        # start restAPI
        log = logging.getLogger('werkzeug')
        log.setLevel(logging.ERROR)
        self.collectT = threading.Thread(
            target=lambda: self.app.run(host='localhost', port=4998, debug=False, use_reloader=False)
        )
        self.collectT.setDaemon(True)
        self.collectT.start()

        if os.path.exists(os.path.join(self.current_dir, "fork.csv")):
            os.remove(os.path.join(self.current_dir, "fork.csv"))
        if os.path.exists(os.path.join(self.current_dir, "start.csv")):
            os.remove(os.path.join(self.current_dir, "start.csv"))
        # remove tmp.csv (if any)
        if os.path.exists(os.path.join(self.current_dir, "tmp.csv")):
            os.remove(os.path.join(self.current_dir, "tmp.csv"))

        optstr = ",".join(["%s=%s" % (k, v) for k, v in options.items()])
        # restrict OL in a cgroup
        cg_name = "ol"
        cgroup_path = f"/sys/fs/cgroup/{cg_name}"
        if os.path.exists(cgroup_path):
            os.rmdir(cgroup_path)
        os.makedirs(cgroup_path, exist_ok=True)
        cmd = ["sudo", "cgexec", "-g", f"memory,cpu:{cg_name}", './ol', 'worker', 'up', '-d']
        if optstr:
            cmd.extend(['-o', optstr])
        print(cmd)
        out = subprocess.check_output(cmd, cwd=self.ol_dir)
        print(str(out, 'utf-8'))

        if options.get("features.warmup", False):
            self.warmup_mem = get_total_mem(self.config["cg_dir"], MODE="PSS")

        match = re.search(r"PID: (\d+)", str(out, 'utf-8'))
        if match:
            pid = match.group(1)
            self.pid = pid
            print(f"The OL PID is {pid}")
            return 0
        else:
            print("No PID found in the text.")
            return -1

    def kill_worker(self, options={}):
        if self.collectT:
            write_list_to_file(os.path.join(self.current_dir, "fork.csv"), self.fork_list)
            write_list_to_file(os.path.join(self.current_dir, "start.csv"), self.start_list)
            self.fork_list = []
            self.start_list = []

        if options.get("save_metrics", False):
            write_list_to_file(os.path.join(self.current_dir, "tmp.csv"), self.metrics)
            csv_name = options.get("csv_name", "metrics.csv")
            if os.path.exists(csv_name):
                os.remove(csv_name)
            os.rename(os.path.join(self.current_dir, "tmp.csv"),  csv_name)
        self.metrics = []

        if not self.pid:
            print("PID has not been set")
            return -1
        try:
            cmd = ['./ol', 'worker', 'down']
            out = subprocess.check_output(cmd, cwd=self.ol_dir)
            print(str(out, 'utf-8'))
        except Exception as e:
            print(e)
            print("force kill")

            print(f"Killing process {self.pid} on port 5000")
            subprocess.run(['kill', '-9', self.pid], cwd=self.ol_dir)

            cmd = ['./ol', 'worker', 'force-cleanup']
            subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, cwd=self.ol_dir)

            process = subprocess.Popen(['./ol', 'worker', 'up'], cwd=self.ol_dir)
            os.kill(process.pid, signal.SIGINT)

            cmd = ['./ol', 'worker', 'force-cleanup']
            subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, cwd=self.ol_dir)
        return 0

    def deploy_func(self, func_config):
        func_path = f"{self.ol_dir}default-ol/registry/{func_config['name']}"

        if os.path.exists(func_path):
            shutil.rmtree(func_path)
        os.makedirs(func_path, exist_ok=True)

        with open(os.path.join(func_path, "f.py"), 'w') as f:
            f.write(func_config["code"])
        with open(os.path.join(func_path, "requirements.txt"), 'w') as f:
            f.write(func_config["requirements_txt"])

    # use go collector
    def invoke_func(self, func_name, options={}):
        if not options:
            url = f"http://localhost:5000/run/{func_name}"
            resp = requests.post(url)
        else:
            if "url" in options.keys() and options["url"] != "":
                url = options["url"]
            else:
                url = f"http://localhost:5000/run/{func_name}"
            resp = requests.post(url, json=options["req_body"])

        if resp.status_code != 200:
            raise Exception(f"Request to {url} failed, resp: {resp.text}, status code: {resp.status_code}")

        try:
            resp_body = resp.json()
        except Exception as e:
            raise Exception(f"Error parsing response: {e}, the response is {resp.text}")
        if isinstance(resp_body, dict):
            resp_body["received"] = time.time() * 1000
            with metrics_lock:
               record = {
                    "invoke_id": resp_body["invoke_id"],
                    "req": resp_body["req"],
                    "received": time.time() * 1000,
                    "start_pullHandler": resp_body["start_pullHandler"],
                    "end_pullHandler": resp_body["end_pullHandler"],
                    "start_create": resp_body["start_create"] if "start_create" in resp_body else 0,
                    "end_create": resp_body["end_create"] if "end_create" in resp_body else 0,
                    "start_import": resp_body["start_import"],
                    "end_import": resp_body["end_import"],
                    "start_execute": resp_body["start_execute"],
                    "end_execute": resp_body["end_execute"],
                    "failed": resp_body["failed"],
                    "split_generation": resp_body["split_gen"],
                    "zygote_miss": resp_body["zygote_miss"] if "zygote_miss" in resp_body else 0,
                }
            self.metrics.append(record)
        resp.close()
        return resp_body, None

def write_list_to_file(file_path, list_data):
    mode = 'a' if os.path.exists(file_path) else 'w'

    s = time.time()
    with open(file_path, mode=mode, newline='', encoding='utf-8') as file:
        writer = csv.writer(file)
        if mode == 'w':
            writer.writerow(list_data[0].keys())
        for item in list_data:
            writer.writerow(item.values())
    e = time.time()
    print(f"write to file {file_path} takes {e-s} seconds")

File: openlambda/ol.go
package openlambda

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	platform "platform_adapter_go"
	"reflect"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
)

type LatencyRecord struct {
	Name             string   `json:"name"`
	SplitGen         int      `json:"split_gen"`
	Req              float64  `json:"req"`
	Received         float64  `json:"received"`
	StartCreate      float64  `json:"start_create"`
	EndCreate        float64  `json:"end_create"`
	StartPullHandler float64  `json:"start_pullHandler"`
	EndPullHandler   float64  `json:"end_pullHandler"`
	Unpause          float64  `json:"unpause"`
	StartImport      float64  `json:"start_import"`
	EndImport        float64  `json:"end_import"`
	StartExecute     float64  `json:"start_execute"`
	EndExecute       float64  `json:"end_execute"`
	ZygoteMiss       int      `json:"zygote_miss"`
	SbID             string   `json:"sb_id"`
	Failed           []string `json:"failed"`
}

func (record *LatencyRecord) toSlice() []string {
	failedStr := "[]"
	if len(record.Failed) > 0 {
		failedStr = fmt.Sprintf("[%s]", strings.Join(record.Failed, ","))
	}
	return []string{
		record.Name,
		strconv.Itoa(record.SplitGen),
		fmt.Sprintf("%.3f", record.Req),
		fmt.Sprintf("%.3f", record.Received),
		fmt.Sprintf("%.3f", record.StartCreate),
		fmt.Sprintf("%.3f", record.EndCreate),
		fmt.Sprintf("%.3f", record.StartPullHandler),
		fmt.Sprintf("%.3f", record.EndPullHandler),
		fmt.Sprintf("%.3f", record.Unpause),
		fmt.Sprintf("%.3f", record.StartImport),
		fmt.Sprintf("%.3f", record.EndImport),
		fmt.Sprintf("%.3f", record.StartExecute),
		fmt.Sprintf("%.3f", record.EndExecute),
		strconv.Itoa(record.ZygoteMiss),
		record.SbID,
		failedStr,
	}
}

func (record *LatencyRecord) getHeaders() []string {
	return []string{"name", "split_gen", "req", "received", "start_create", "end_create", "start_pullHandler", "end_pullHandler",
		"unpause", "start_import", "end_import", "start_execute", "end_execute",
		"zygote_miss", "sb_id", "failed"}
}

func (record *LatencyRecord) parseJSON(jsonData []byte) error {
	//println(string(jsonData))
	err := json.Unmarshal(jsonData, &record)
	if err != nil {
		return err
	}

	// Use reflection to ensure fields() and fieldByName() are available
	v := reflect.ValueOf(record).Elem()
	for i := 0; i < v.NumField(); i++ {
		field := v.Field(i)

		// Only check for nil values for pointers, slices, and maps
		if field.Kind() == reflect.Ptr || field.Kind() == reflect.Slice || field.Kind() == reflect.Map {
			if field.IsNil() {
				switch field.Kind() {
				case reflect.Int:
					field.SetInt(0)
				case reflect.Slice:
					field.Set(reflect.MakeSlice(field.Type(), 0, 0))
				}
			}
		}
	}
	return nil
}

type OpenLambda struct {
	platform.BasePlatformAdapter
	PID            int
	olDir          string
	runUrl         string
	collectLatency bool
	latencyRecords []LatencyRecord
	LatenciesMutex *sync.Mutex

	currentDir string
}

func (o *OpenLambda) StartWorker(options map[string]interface{}) error {
	tmpFilePath := o.currentDir + "/tmp.csv"
	if _, err := os.Stat(tmpFilePath); err == nil {
		os.Remove(tmpFilePath)
	}

	var optParts []string
	for k, v := range options {
		optParts = append(optParts, k+"="+v.(string))
	}
	optstr := strings.Join(optParts, ",")

	cgName := "ol"
	cgroupPath := "/sys/fs/cgroup/" + cgName
	if _, err := os.Stat(cgroupPath); err == nil {
		os.Remove(cgroupPath)
	}
	os.MkdirAll(cgroupPath, 0755)

	cmdArgs := []string{"cgexec", "-g", "memory,cpu:" + cgName, "./ol", "worker", "up", "-d"}
	if optstr != "" {
		cmdArgs = append(cmdArgs, "-o", optstr)
	}
	cmd := exec.Command("sudo", cmdArgs...)
	cmd.Dir = o.olDir
	out, err := cmd.Output()
	if err != nil {
		return fmt.Errorf("failed to start worker: %v", err)
	}

	output := string(out)

	// todo: warmup and profiling

	re := regexp.MustCompile(`PID: (\d+)`)
	match := re.FindStringSubmatch(output)
	if len(match) > 1 {
		pid, err := strconv.Atoi(match[1])
		if err != nil {

		}
		o.PID = pid
		fmt.Println("The OL PID is", pid)
		return nil
	} else {
		return fmt.Errorf("failed to parse PID from output: %s", output)
	}
}

func (o *OpenLambda) KillWorker(options map[string]interface{}) error {
	// kill worker
	if o.PID == 0 {
		fmt.Println("PID has not been set")
		return fmt.Errorf("PID has not been set")
	}

	cmd := exec.Command("./ol", "worker", "down")
	cmd.Dir = o.olDir
	_, err := cmd.Output()
	if err != nil {
		fmt.Println(err)
		fmt.Println("force kill")

		fmt.Printf("Killing process %d on port 5000\n", o.PID)
		killCmd := exec.Command("kill", "-9", fmt.Sprint(o.PID))
		killCmd.Dir = o.olDir
		killCmd.Run()

		cleanupCmd := exec.Command("./ol", "worker", "force-cleanup")
		cleanupCmd.Dir = o.olDir
		cleanupCmd.Stdout = nil
		cleanupCmd.Stderr = nil
		cleanupCmd.Run()

		upCmd := exec.Command("./ol", "worker", "up")
		upCmd.Dir = o.olDir
		if err := upCmd.Start(); err != nil {
			fmt.Println("Failed to start the command:", err)
		} else {
			// Send SIGINT to the process
			if err := upCmd.Process.Signal(syscall.SIGINT); err != nil {
				fmt.Println("Failed to send SIGINT:", err)
			}
		}

		cleanupCmd2 := exec.Command("./ol", "worker", "force-cleanup")
		cleanupCmd2.Dir = o.olDir
		cleanupCmd2.Stdout = nil
		cleanupCmd2.Stderr = nil
		cleanupCmd2.Run()
		fmt.Printf("force kill done\n")
		return nil
	} else {
		return nil
	}
}

func (o *OpenLambda) DeployFunc(f platform.Function) error {
	// write code to registry dir
	meta := f.Meta
	path := fmt.Sprintf(o.olDir+"/default-ol/registry/%s", f.Name)
	if os.IsExist(os.MkdirAll(path, 0777)) {
		err := os.RemoveAll(path)
		if err != nil {
			panic(err)
		}
		err = os.MkdirAll(path, 0777)
		if err != nil {
			panic(err)
		}
	}

	_lines := f.Code
	var lines []string
	for _, line := range _lines {
		lines = append(lines, line)
	}
	code := strings.Join(lines, "\n")

	funcPath := filepath.Join(path, "f.py")
	requirementsInPath := filepath.Join(path, "requirements.in")
	requirementsTxtPath := filepath.Join(path, "requirements.txt")

	if err := ioutil.WriteFile(funcPath, []byte(code), 0777); err != nil {
		panic(err)
	}
	if err := ioutil.WriteFile(requirementsInPath, []byte(meta.RequirementsIn), 0777); err != nil {
		panic(err)
	}
	if err := ioutil.WriteFile(requirementsTxtPath, []byte(meta.RequirementsTxt), 0777); err != nil {
		panic(err)
	}

	return nil
}

func (o *OpenLambda) InvokeFunc(funcName string, timeout int, options map[string]interface{}) error {
	// invoke function
	url := o.runUrl + funcName
	var resp *http.Response
	var err error

	jsonData, err := json.Marshal(options)
	if err != nil {
		log.Fatalf("failed to marshal latency dict: %v", err)
		return err
	}
	client := &http.Client{
		Timeout: time.Duration(timeout) * time.Second,
	}
	resp, err = client.Post(url, "text/json", bytes.NewBuffer(jsonData))
	if err != nil {
		log.Fatalf("failed to post to %s: %v", url, err)
		return err
	}
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatalf("failed to read response body: %v", err)
		return err
	}
	resp.Body.Close()

	if o.collectLatency {
		var record LatencyRecord
		err = record.parseJSON(body)
		if err != nil {
			log.Fatalf("failed to parse latency record: %v", err)
			return err
		}
		o.LatenciesMutex.Lock()
		o.latencyRecords = append(o.latencyRecords, record)
		o.LatenciesMutex.Unlock()
	}
	return nil
}

func NewOpenLambda() *OpenLambda {
	return &OpenLambda{}
}
